### PR TITLE
BDOG-1728 - Track Exemptions

### PR DESCRIPTION
**Pull request by hmrc-phil**

### CODE REVIEW

File: app/uk/gov/hmrc/leakdetection/model/Leak.scala [hunk 1]
```diff
@@ -37,7 +37,8 @@ case class Leak( repoName   : String,
                  urlToSource: String,
                  lineText   : String,
                  matches    : List[Match],
-                 priority   : String)
+                 priority   : String,
+                 excluded   : Boolean)
```
Comment (colin-lamed) [inline app/uk/gov/hmrc/leakdetection/model/Leak.scala:41]: nice to give booleans an `is` prefix

File: app/uk/gov/hmrc/leakdetection/model/Report.scala [hunk 2]
```diff
@@ -62,7 +62,9 @@ final case class Report(
                          timestamp        : Instant,
                          author           : String,
                          totalLeaks       : Int,
-                         rulesViolated    : Map[String, Int]
+                         totalWarnings    : Int = 0,
+                         rulesViolated    : Map[String, Int],
+                         exclusions       : Map[String, Int]
```
Comment (colin-lamed) [inline app/uk/gov/hmrc/leakdetection/model/Report.scala:67]: what does the key represent, `RuleId`? Maybe introduce a value class?

File: app/uk/gov/hmrc/leakdetection/model/Leak.scala (hunk 1 above)
Comment (tomwadeson) [inline app/uk/gov/hmrc/leakdetection/model/Leak.scala:41]: `exempt`?

File: test/uk/gov/hmrc/leakdetection/scanner/RegexMatchingEngineSpec.scala [hunk 3]
```diff
@@ -18,13 +18,14 @@ package uk.gov.hmrc.leakdetection.scanner
...
+import org.scalatest.wordspec.AnyWordSpec
 import uk.gov.hmrc.leakdetection.config.Rule
+import uk.gov.hmrc.leakdetection.config.Rule.Priority
 
-class RegexMatchingEngineSpec extends AnyFreeSpec with MockitoSugar with Matchers {
+class RegexMatchingEngineSpec extends AnyWordSpec with MockitoSugar with Matchers {
 
-  "run" - {
+  "run" should {
     "should scan all the files in all subdirectories and return a report with correct file paths" in {
```
Comment (colin-lamed) [inline test/uk/gov/hmrc/leakdetection/scanner/RegexMatchingEngineSpec.scala:29]: double `should` now?

File: app/uk/gov/hmrc/leakdetection/scanner/RegexScanner.scala [hunk 4]
```diff
@@ -67,12 +67,30 @@ case class RegexScanner(rule: Rule, lineLengthLimit: Int, lineExemptions: Seq[St
...
           )
         )
       case _ => None
     }
 
+  val fileExtensionR = """\.[A-Za-z0-9]+$""".r
+  private def isLineExempt(ruleId: String, filePath: String, line: String, inLineExemption: Boolean, ruleExemptions: Seq[RuleExemption]): Boolean = {
+      inLineExemption ||
+      ruleExemptions
+        .filter(_.ruleId == ruleId)
```
Comment (tomwadeson) [inline app/uk/gov/hmrc/leakdetection/scanner/RegexScanner.scala:83]: Smash the two filters together?

Comment (tomwadeson) [inline app/uk/gov/hmrc/leakdetection/scanner/RegexScanner.scala:83]: Smash the two filters together?

File: app/uk/gov/hmrc/leakdetection/scanner/RegexScanner.scala [hunk 5]
```diff
@@ -67,12 +67,30 @@ case class RegexScanner(rule: Rule, lineLengthLimit: Int, lineExemptions: Seq[St
...
+        .filter(_.filePaths.contains(filePath))
+        .flatMap(_.text)
+        .exists(line.contains(_))
+  }
+
+  private def isFileExempt(ruleId: String, filePath: String, ruleExemptions: Seq[RuleExemption]): Boolean = {
+    ruleExemptions
+      .filter(_.ruleId == ruleId)
```
Comment (tomwadeson) [inline app/uk/gov/hmrc/leakdetection/scanner/RegexScanner.scala:91]: As above

Comment (tomwadeson) [inline app/uk/gov/hmrc/leakdetection/scanner/RegexScanner.scala:91]: As above

**Review by tomwadeson — DISMISSED** [review]

**Review by colin-lamed — DISMISSED** [review]

**Review by tomwadeson — APPROVED** [review]

**Review by colin-lamed — APPROVED** [review]

**Pull request merged**